src/server/dbconn.cpp

```diff
@@ -0,0 +1,396 @@
+		account_ = mariadb::account::create(c["db_host"].str(), c["db_user"].str(), c["db_password"].str());
+		account_->set_connect_option(mysql_option::MYSQL_SET_CHARSET_NAME, std::string("utf8mb4"));
+		account_->set_schema(c["db_name"].str());
+		connection_ = mariadb::connection::create(account_);
+	}
+	catch(const mariadb::exception::base& e)
+	{
+		log_sql_exception("Failed to connect to the database!", e);
+	}
+}
+
+void dbconn::log_sql_exception(const std::string& text, const mariadb::exception::base& e)
+{
+	ERR_SQL << text << '\n'
+			<< "what: " << e.what() << '\n'
+			<< "error id: " << e.error_id() << std::endl;
+}
+
+//
+// queries
+//
+std::string dbconn::get_uuid()
+{
+	try
+	{
+		return get_single_string("SELECT UUID()");
+	}
+	catch(const mariadb::exception::base& e)
+	{
+		log_sql_exception("Could not retrieve a UUID!", e);
+		return "";
+	}
+}
+
+std::string dbconn::get_tournaments()
+{
+	if(db_tournament_query_ == "")
+	{
+		return "";
+	}
+
+	try
+	{
+		tournaments t;
+		get_complex_results(t, db_tournament_query_);
+		return t.str();
+	}
+	catch(const mariadb::exception::base& e)
+	{
+		log_sql_exception("Could not retrieve the tournaments!", e);
+		return "";
+	}
+}
+
+bool dbconn::user_exists(const std::string& name)
+{
+	try
+	{
+		return exists("SELECT 1 FROM `"+db_users_table_+"` WHERE UPPER(username)=UPPER(?)", name);
+	}
+	catch(const mariadb::exception::base& e)
+	{
+		log_sql_exception("Unable to check if user row for '"+name+"' exists!", e);
+		return false;
+	}
+}
+
+bool dbconn::extra_row_exists(const std::string& name)
+{
+	try
+	{
+		return exists("SELECT 1 FROM `"+db_extra_table_+"` WHERE UPPER(username)=UPPER(?)", name);
+	}
+	catch(const mariadb::exception::base& e)
+	{
+		log_sql_exception("Unable to check if extra row for '"+name+"' exists!", e);
+		return false;
+	}
+}
+
+bool dbconn::is_user_in_group(const std::string& name, int group_id)
+{
+	try
+	{
+		return exists("SELECT 1 FROM `"+db_users_table_+"` u, `"+db_user_group_table_+"` ug WHERE UPPER(u.username)=UPPER(?) AND u.USER_ID = ug.USER_ID AND ug.GROUP_ID = ?", name, group_id);
+	}
+	catch(const mariadb::exception::base& e)
+	{
+		log_sql_exception("Unable to check if the user '"+name+"' is in group '"+std::to_string(group_id)+"'!", e);
+		return false;
+	}
+}
+
+ban_check dbconn::get_ban_info(const std::string& name, const std::string& ip)
+{
+	try
+	{
+		ban_check b;
+		get_complex_results(b, "select ban_userid, ban_email, case when ban_ip != '' then 1 when ban_userid != 0 then 2 when ban_email != '' then 3 end as ban_type, ban_end from `"+db_banlist_table_+"` where (ban_ip = ? or ban_userid = (select user_id from `"+db_users_table_+"` where UPPER(username) = UPPER(?)) or UPPER(ban_email) = (select UPPER(user_email) from `"+db_users_table_+"` where UPPER(username) = UPPER(?))) AND ban_exclude = 0 AND (ban_end = 0 OR ban_end >= ?)", ip, name, name, std::time(nullptr));
```

Comment:
I suggest to add a comment like: `the ban_type part of this query needs to stay in sync with the user_handler::BAN_TYPE enum`

Reply:
Updated.